### ISSUE TITLE
Synthetic pattern-based integration test for generation correctness (single-chunk + rolling)

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/MEDS_EIC_AR#91]
## Summary

Replace the current "does the CLI produce non-empty parquet?" smoke-test style with a real **generation-correctness** integration test: train a tiny model on CPU against a synthetic pattern dataset where we know the ground truth, then assert that generation — both single-chunk and rolling — reproduces the patterns the model was trained on. The point is to validate not just that the plumbing wires up but that the model + generation path actually **works** end-to-end on something we can verify.

**Not for this PR.** Design-only for now; implementation as a follow-up in dev after #86 lands.

## Motivation

Our current tests fall into two camps:

- **Unit / doctest level**: `Model.generate`, `_rolling_generate`, `_truncate_at_first_eos` are exercised on random-init models with deterministic seeds. The doctests prove the **code path** executes and the **shapes** are right. They don't prove that, if the model actually learned something, generation would recover it.
- **CLI integration level**: `test_generate_trajectories_runs` / `test_generate_trajectories_rolling_runs` drive the full `MEICAR_generate_trajectories` CLI and check that parquet files exist, splits line up, and — as of #86's review follow-up — that the rolling path produces output longer than the single-chunk path could. Better than pure existence checks, but still silent on whether the generated content is *right*.

The missing middle: a test that (1) trains a model quickly enough to fit in CI, (2) on a dataset simple enough that correct generation is unambiguous, and (3) exercises the **full generation path** so we catch regressions that break the generation math without breaking any shape or plumbing invariant.

### The pattern dataset

Define a small probabilistic grammar over a tiny vocabulary. Each "program" is one of a few recognizable fixed substrings, emitted with stochastic interleaving:

```
program_A := "A B C D"     # weight 0.4
program_B := "1 2 3"       # weight 0.3
program_C := "R S T U V W X Y Z"  # weight 0.3

sequence := [program] ("|" [program])* EOS
```

The separator `"|"` between programs lets the model learn "after |, pick any program" while still being able to complete a program once it's started. Vocab size: ~20 tokens (A-D, 1-3, R-Z, `|`, `BOS`, `EOS`, plus a pad). Each program is short enough that completing it is a strong test of one-step lookahead.

**Why this shape of task:**
- **Correct generation is unambiguously verifiable.** After seeing `A B`, the model must output `C` — there's no ambiguity, no distributional metric needed. Grammar adherence is a binary per-step check.
- **Small enough to train fast.** ~20 tokens of vocab, ~2 layers, hidden dim ~32, max_seq_len ~64 should train in a few minutes on CPU to near-perfect grammar adherence.
- **Long enough to exercise rolling.** If we set `max_seq_len = 16` and generate ~50 tokens, the rolling loop must cross several chunks, and we can check that grammar adherence holds through sliding boundaries.

### Training corpus

- Generate ~5k–20k training sequences from the grammar.
- Each sequence: 5–15 programs concatenated with `|` separators, ending in `EOS`, total length capped at `max_seq_len` or slightly longer to force truncation during training.
- Save as a minimal MEDS-compatible format (or a custom synthetic datamodule that bypasses MEDS — TBD, see the "Where does this live" section).

### Model config

- `num_hidden_layers=2`, `num_attention_heads=2`, `hidden_size=32`, `max_seq_len=16` or `32`.
- Train for ~1–5 epochs on CPU; target is not accuracy per se, just "has converged on the grammar." A handful of minutes tops.

### Training budget and CI fit

- Target wall-clock: **under 60 seconds** for training + generation + assertion on a CI runner. If that proves optimistic, 2 minutes is still acceptable for a single integration test; anything longer needs scoping down (smaller corpus, fewer epochs, smaller model).
- This is a **session-scoped** fixture — train once per test session, reuse across generation tests.

### Assertions

Two test cases, both on the trained model:

**1. Single-chunk generation reproduces the grammar.**
- Seed the model with `BOS A`, call `Model.generate()` without rolling.
- Assert: the continuation starts with `B C D` and then either `|` or `EOS`.
- Repeat for each program prefix (`BOS 1` → `2 3 ...`, `BOS R` → `S T U V W X Y Z ...`).
- Repeat with mid-program prefixes (`BOS A B` → `C D ...`).

**2. Rolling generation preserves grammar across sliding boundaries.**
- Seed the model with `BOS` and call `Model.generate(max_new_tokens=80)` with rolling enabled. Since `max_seq_len=16`, this forces the loop to cross at least 5 sliding boundaries.
- Assert: every program that appears in the output is fully formed (no `A B C` without a following `D`, etc.) or, if truncated, the truncation happens cleanly at a program boundary.
- Walk the generated sequence token by token, tracking "am I inside a program and which one?" — assert that every transition is grammar-valid.

**The rolling assertion is the real payoff.** If rolling generation corrupts the model's learned patterns at sliding boundaries (e.g., because of position reset affecting learned behaviors, or because the sliding window drops context the model was relying on), we'll see grammar violations. If rolling generation is a faithful extension of single-chunk generation, grammar adherence should be indistinguishable.

## Where does this live

Two options, with different tradeoffs:

**Option A: synthetic MEDS dataset.** Build the pattern grammar as a MEDS-formatted dataset (with `subject_id`, `time`, `code`, etc. columns) so we can drive it through the existing datamodule, pretraining CLI, and generation CLI. **Pros:** exercises the same pipeline users hit. **Cons:** more ceremony — MEDS schema demands certain fields, we'd have to fake timestamps and prediction windows, and the grammar has to be bolted onto the MEDS-code-vocabulary conventions.

**Option B: direct `Model` instantiation with a custom synthetic datamodule.** Skip MEDS entirely — define our own `torch.utils.data.Dataset` subclass that yields grammar-generated integer sequences, train `Model` directly in a small training loop, then call `Model.generate` directly. **Pros:** minimal ceremony, fastest iteration, directly tests the `Model` API. **Cons:** bypasses the MEDS datamodule and CLI plumbing, which are themselves worth integration-testing — though those are already covered by the existing `test_generate_trajectories_runs` tests.

**Recommendation:** start with **Option B** for speed and iteration cost. The existing CLI tests already cover the datamodule/CLI plumbing with real-ish MEDS data; this synthetic test is specifically about generation-correctness of the model + generation path, not the CLI. If Option B proves valuable and we want to extend it to also cover the CLI plumbing, a follow-up can port the grammar to a MEDS-compatible schema.

## Open questions

- **Is CPU training fast enough?** Needs an empirical check with the proposed model size + corpus size. If yes, full integration. If no, we may need to pre-train the synthetic model and commit the checkpoint, which is awkward.
- **How strict should grammar assertions be?** A too-strict assertion (e.g., "100% of programs are perfectly formed") may flake on under-trained runs. A looser assertion (e.g., "≥ 90% grammar adherence") is probably the right target, with the threshold tuned empirically.
- **Should this test gate the PR**, or run as a slow / separate job? If wall-clock is ~60s, gate it. If it's multiple minutes, slow-marker is fine.
- **Does this test replace the existing CLI-driven integration tests, or supplement them?** Probably supplement — the CLI tests cover different risks (plumbing, config, datamodule). This one covers generation correctness specifically.

## Scope for the implementation PR

- [ ] Prototype the grammar generator (pure Python, no deps).
- [ ] Build a minimal synthetic datamodule (Option B) and training loop, or wire the grammar into a MEDS-compatible schema (Option A).
- [ ] Train a tiny model to grammar adherence, measure wall-clock.
- [ ] Write the two assertions (single-chunk and rolling) and validate they pass on a trained model and **fail** on a random-init model (negative control).
- [ ] Decide on Option A vs. B based on wall-clock and ceremony cost.
- [ ] Land the test alongside documentation in the test file explaining what it validates and why.

## Out of scope (for the implementation PR)

- Testing SCOPE / REACH or any advanced generation methods (#87).
- Benchmarking or replacing the rolling loop with `transformers-community/sink_cache` (#90).
- Extending the grammar to include numerical features, time deltas, or other MEDS-specific constructs — keep the initial version minimal and expand later if useful.

Tracking here because @mmcdermott and I agreed in the #86 / #82 discussion that this is the right next-step test to have before we start adding more complex generation features. Belongs in the next dev PR after #86 merges.

**Comment by mmcdermott** [comment]
Implemented in PR #93, which added `tests/test_pattern_generation.py` with a CPU-trained model on a synthetic grammar, GrammarFSM validation, negative controls via ALT_PROGRAMS, and both single-chunk and rolling generation-correctness tests. Merged to dev and then main via PR #94.
